### PR TITLE
Layer refactor

### PR DESCRIPTION
**Pull request by IagoLast**

### CODE REVIEW

File: src/api/layer.js [hunk 1]
```diff
@@ -125,24 +125,15 @@ export default class Layer {
...
-        source.bindLayer(this._onDataframeAdded.bind(this), this._onDataFrameRemoved.bind(this), this._onDataLoaded.bind(this));
-        if (this._source !== source) {
-            this._freeSource();
+    setSource(source) {
+        if (this._source === source) {
+            return;
         }
+        this._checkSource(source);
+        this._freeSource();
         this._source = source;
```
Comment (davidmanzanares) [inline src/api/layer.js]: This refactor looks very, very, dangerous to me. You are setting the object source to the new source before waiting for checking the metadata against the style.

File: src/api/layer.js [hunk 2]
```diff
@@ -157,16 +148,11 @@ export default class Layer {
...
-                this._style.onChange(null);
-            }
-            this._style = style;
-            this._style.onChange(() => {
-                this._styleChanged(style);
-            });
-            return r;
-        });
+        this._style.onChange(null);
+        this._style = style;
```
Comment (davidmanzanares) [inline src/api/layer.js]: Similar problem here, this sets the style without checking that the style is valid for the source

File: src/api/layer.js [hunk 3]
```diff
@@ -157,16 +148,11 @@ export default class Layer {
...
-            if (this._style) {
-                this._style.onChange(null);
-            }
-            this._style = style;
-            this._style.onChange(() => {
-                this._styleChanged(style);
-            });
-            return r;
-        });
+        this._style.onChange(null);
```
Comment (davidmanzanares) [inline src/api/layer.js]: What if there is no style?

File: src/api/layer.js [hunk 4]
```diff
@@ -194,25 +180,21 @@ export default class Layer {
...
-            if (this._style) {
-                this._style.onChange(null);
-            }
-            this._style = style;
-            this._style.onChange(() => {
-                this._styleChanged(style);
-            });
-            return r;
-        });
+        this._style = style;
```
Comment (davidmanzanares) [inline src/api/layer.js]: Similar problem

File: src/api/layer.js (hunk 1 above)
Comment (IagoLast) [reply src/api/layer.js]: Whats the possible problem?

Comment (davidmanzanares) [reply src/api/layer.js]: Leaving the layer object in an invalid state.

File: src/api/layer.js (hunk 3 above)
Comment (IagoLast) [reply src/api/layer.js]: It fails but it's impossible since the constructor forces you to set a source and a style

Comment (davidmanzanares) [reply src/api/layer.js]: But the constructor set it asynchronously, right?

File: examples/examples.json [hunk 5]
```diff
@@ -16,11 +16,6 @@
                     "desc": "Add multiple CARTO layers to your map",
                     "file": "basics/multi-layer.html"
                 },
-                {  
-                    "title": "Change the source",
```
Comment (davidmanzanares) [inline examples/examples.json]: Why do you remove two examples?

File: src/api/layer.js [hunk 6]
```diff
@@ -179,18 +127,15 @@ export default class Layer {
...
-            });
-            return r;
-        });
+
+        await this._styleChanged(style);
+        if (this._style) {
+            this._style.onChange(null);
+        }
+        this._style = style;
+        this._style.onChange(this._styleChanged.bind(this));
```
Comment (davidmanzanares) [inline src/api/layer.js]: `_styleChanged` doesn't receive a style parameter?

File: src/api/layer.js [hunk 7]
```diff
@@ -225,35 +170,36 @@ export default class Layer {
...
+        this._styleChanged(style).then(() => {
             if (this._style) {
                 this._style.onChange(null);
             }
             this._style = style;
-            this._style.onChange(() => {
-                this._styleChanged(style);
-            });
-            return r;
+            this._style.onChange(this._styleChanged.bind(this));
```
Comment (davidmanzanares) [inline src/api/layer.js]: same

File: examples/examples.json (hunk 5 above)
Comment (IagoLast) [reply examples/examples.json]: Because `setSource` method was removed

File: src/api/layer.js (hunk 6 above)
Comment (IagoLast) [reply src/api/layer.js]: yes, the `style` object will call this method passing `this` as parameter

File: examples/examples.json (hunk 5 above)
Comment (davidmanzanares) [reply examples/examples.json]: Yes, but, the functionality is still there, we should use `update()` instead

File: src/api/layer.js (hunk 6 above)
Comment (davidmanzanares) [reply src/api/layer.js]: Ok

**Comment by Jesus89** [comment]
Are you agree on removing `setSource` function?

Nice code
```js
citiesLayer.setSource(allCitiesSource);
```

won't be nice anymore
```js
citiesLayer.update(allCitiesSource, citiesLayer.getStyle());
```

**Comment by davidmanzanares** [comment]
@Jesus89 I agree. We can always add it later. This new form is uglier, but it is safer and its semantics are easier to understand. Also, we discourage the use of SQL, particularly for simple things that can be done by the style, like filtering.

In the future, I wouldn't be surprised if we made the second argument of `update()` optional, though.

**Comment by IagoLast** [comment]
@davidmanzanares can we merge this?

**Comment by davidmanzanares** [comment]
Yes :rocket:

**Pull request merged**